daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/PreprocessorSpec.scala

```diff
@@ -68,6 +52,129 @@ class PreprocessorSpec extends AnyWordSpec with Inside with Matchers {
+
+        intermediaryResult shouldBe a[ResultNeedPackage[_]]
+
+        val finalResult = intermediaryResult.consume(_ => None, pkgs.get, _ => None)
+
+        inside(finalResult) { case Right(disclosedContracts) =>
+          all(disclosedContracts.toList) should acceptDisclosedContract
+        }
+      }
+    }
+  }
+}
+
+object PreprocessorSpec {
+
+  implicit val defaultPackageId: Ref.PackageId = defaultParserParameters.defaultPackageId
+
+  val pkg =
+    p"""
+        module Mod {
+
+          record @serializable Record = { owners: List Party, data : Int64 };
+
+          template (this : Record) = {
+            precondition True;
+            signatories Mod:Record {owners} this;
+            observers Mod:Record {owners} this;
+            agreement "Agreement";
+            key @(List Party) (Mod:Record {owners} this) (\ (parties: List Party) -> parties);
+          };
+
+        }
+    """
+  val pkgs = Map(defaultPackageId -> pkg)
+  val alice: Party = Ref.Party.assertFromString("Alice")
+  val parties: ValueList = ValueList(FrontStack(ValueParty(alice)))
+  val testKeyName: String = "test-key"
+  val contractId: ContractId =
+    Value.ContractId.V1.assertBuild(
+      crypto.Hash.hashPrivateKey("test-contract-id"),
+      Bytes.assertFromString("deadbeef"),
+    )
+  val templateId: Ref.Identifier = Ref.Identifier.assertFromString("-pkgId-:Mod:Record")
+  val templateType: Ref.TypeConName = Ref.TypeConName.assertFromString("-pkgId-:Mod:Record")
+  val key: Value.ValueRecord = Value.ValueRecord(
+    None,
+    ImmArray(
+      None -> Value.ValueText(testKeyName),
+      None -> Value.ValueList(FrontStack.from(ImmArray(ValueParty(alice)))),
+    ),
+  )
+  val keyHash: Hash = crypto.Hash.assertHashContractKey(templateType, key)
+  val normalizedContract: command.DisclosedContract =
+    buildDisclosedContract(keyHash, withNormalization = true)
+  val nonNormalizedContract: command.DisclosedContract =
+    buildDisclosedContract(keyHash, withNormalization = false)
+
+  def buildDisclosedContract(
+      keyHash: Hash,
+      withNormalization: Boolean,
+  ): command.DisclosedContract = {
+    command.DisclosedContract(
+      templateId,
+      contractId,
+      Value.ValueRecord(
+        if (withNormalization) None else Some(templateId),
+        ImmArray(
+          (if (withNormalization) None else Some(Ref.Name.assertFromString("owners"))) -> parties,
+          (if (withNormalization) None else Some(Ref.Name.assertFromString("data"))) -> Value
+            .ValueInt64(42L),
+        ),
+      ),
+      ContractMetadata(Time.Timestamp.now(), Some(keyHash), ImmArray.Empty),
+    )
   }
 
+  @SuppressWarnings(
+    Array(
+      "org.wartremover.warts.Product",
+      "org.wartremover.warts.Serializable",
+      "org.wartremover.warts.JavaSerializable",
+    )
+  )
+  def acceptDisclosedContract: Matcher[speedy.DisclosedContract] = Matcher { disclosedContract =>
+    val testResult = disclosedContract.argument match {
```

Comment:
is there a reason why you're not build that via `shouldBe`? afaik that would give better errors on failure.

Reply:
I'll refactor the code so that we use an assertion rather than a custom matcher.